lib/agent-redirect.js

```diff
@@ -0,0 +1,209 @@
+const { InvalidArgumentError } = require('./core/errors')
+const util = require('./core/util')
+const EventEmitter = require('events')
+const { getGlobalAgent } = require('./agent')
+const assert = require('assert')
+
+const kMaxRedirections = Symbol('maxRedirections')
+const kAgent = Symbol('agent')
+
+class RedirectHandler {
+  constructor ({ agent, opts, origin, handler }) {
+    this.agent = agent
+    this.location = null
+    this.abort = null
+    this.opts = opts
+    this.origin = origin
+    this.handler = handler
+    this.destroyed = false
+  }
+
+  destroy (err) {
+    this.destroyed = true
+
+    if (this.abort) {
+      this.abort()
+    }
+
+    this.handler.onError(err)
+  }
+
+  get (origin) {
+    return this[kAgent].get(origin)
+  }
+
+  onConnect (abort) {
+    if (this.destroyed) {
+      abort()
+    } else {
+      this.abort = abort
+      this.handler.onConnect(abort)
+    }
+  }
+
+  onUpgrade (statusCode, headers, socket) {
+    this.handler.onUpgrade(statusCode, headers, socket)
+  }
+
+  onError (error) {
+    this.handler.onError(error)
+  }
+
+  onHeaders (statusCode, headers, resume) {
+    if ([300, 301, 302, 303, 307, 308].indexOf(statusCode) === -1) {
+      return this.handler.onHeaders(statusCode, headers, resume)
+    }
+
+    this.location = parseLocation(headers)
+
+    if (!this.location) {
+      return this.handler.onHeaders(statusCode, headers, resume)
+    }
+
+    const { origin, pathname, search } = util.parseURL(new URL(this.location, this.origin))
+    const path = search ? `${pathname || '/'}${search || ''}` : pathname
+
+    // Remove headers referring to the original URL.
+    // By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+    // https://tools.ietf.org/html/rfc7231#section-6.4
+    this.opts.headers = cleanRequestHeaders(this.opts.headers, statusCode === 303)
+    this.opts.path = path
+    this.opts.origin = origin
+
+    // https://tools.ietf.org/html/rfc7231#section-6.4.4
+    // In case of HTTP 303, always replace method to be either HEAD or GET
+    if (statusCode === 303) {
+      this.opts.method = 'GET'
```

Comment:
Little nit here. If the original method was `HEAD`, it must not be changed to get. So the line before should check for `statusCode === 303 && this.opts.method !== 'GET'`